### ISSUE TITLE
"Scope has no PIDs. Refusing" messages when using --cgroup-driver=systemd

### ISSUE THREAD

**Issue by kvaps** [kubernetes/kubernetes#71887]
**What happened**:

I have strange issue when using `systemd` cgroups driver, example messages:

```
systemd[1]: libcontainer-29519-systemd-test-default-dependencies.scope: Scope has no PIDs. Refusing.
systemd[1]: libcontainer-29519-systemd-test-default-dependencies.scope: Scope has no PIDs. Refusing.
systemd[1]: Created slice libcontainer_29519_systemd_test_default.slice.
kubelet[6547]: W1209 03:04:49.649981    6547 container.go:422] Failed to get RecentStats("/libcontainer_29519_systemd_test_default.slice") while determining the next housekeeping: unable to find data in memory cache
systemd[1]: Removed slice libcontainer_29519_systemd_test_default.slice.
systemd[1]: libcontainer-29554-systemd-test-default-dependencies.scope: Scope has no PIDs. Refusing.
systemd[1]: libcontainer-29554-systemd-test-default-dependencies.scope: Scope has no PIDs. Refusing.
systemd[1]: Created slice libcontainer_29554_systemd_test_default.slice.
systemd[1]: Removed slice libcontainer_29554_systemd_test_default.slice.
systemd[1]: libcontainer-29561-systemd-test-default-dependencies.scope: Scope has no PIDs. Refusing.
systemd[1]: libcontainer-29561-systemd-test-default-dependencies.scope: Scope has no PIDs. Refusing.
systemd[1]: Created slice libcontainer_29561_systemd_test_default.slice.
kubelet[6547]: W1209 03:04:50.591527    6547 container.go:523] Failed to update stats for container "/libcontainer_29561_systemd_test_default.slice": open /sys/fs/cgroup/memory/libcontainer_29561_systemd_test_default.slice/memory.use_hierarchy: no such file or directory, continuing to push stats
systemd[1]: Removed slice libcontainer_29561_systemd_test_default.slice.
systemd[1]: libcontainer-29578-systemd-test-default-dependencies.scope: Scope has no PIDs. Refusing.
systemd[1]: libcontainer-29578-systemd-test-default-dependencies.scope: Scope has no PIDs. Refusing.
systemd[1]: Created slice libcontainer_29578_systemd_test_default.slice.
kubelet[6547]: W1209 03:04:50.627114    6547 container.go:422] Failed to get RecentStats("/libcontainer_29578_systemd_test_default.slice") while determining the next housekeeping: unable to find data in memory cache
```

**What you expected to happen**:

No messages

**How to reproduce it (as minimally and precisely as possible)**:

```
# cat /var/lib/kubelet/kubeadm-flags.env
KUBELET_KUBEADM_ARGS=--cgroup-driver=systemd --network-plugin=cni
```

```
# cat /etc/docker/daemon.json 
{
  "exec-opts": ["native.cgroupdriver=systemd"],
  "iptables": false,
  "ip-forward": false,
  "log-driver": "json-file",
  "bridge": "none",
  "log-opts": {
    "max-size": "10m",
    "max-file": "5"
  },
  "storage-driver": "overlay2"
}
```

**Anything else we need to know?**:

When switching `systemd` to `cgroupfs` problem is going out.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T21:04:45Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T20:56:12Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: baremetal
- OS (e.g. from /etc/os-release): `Debian GNU/Linux 9 (stretch)`
- Kernel (e.g. `uname -a`): 
```
Linux pve2 4.15.18-7-pve #1 SMP PVE 4.15.18-26 (Thu, 04 Oct 2018 11:03:06 +0200) x86_64 GNU/Linux
```
- Install tools: `kubeadm`
- Others: `Docker version 18.06.1-ce, build e68fc7a`

/kind bug

**Comment by discostur** [comment]
seeing the same messages like @kvaps 

Docker Version: 18.09.1
Kernel: 3.10.0-957.1.3.el7.x86_64 #1 SMP Thu Nov 29 14:49:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
k8s: GitVersion:"v1.10.11"

**Comment by BSWANG** [comment]
This empty cgroup is created by runc for checking systemd capability:
https://github.com/opencontainers/runc/blob/76520a4bf07b0558aca3b867277d696b41978b44/libcontainer/cgroups/systemd/apply_systemd.go#L109

IMO, Kubelet/cadvisor should ignore this kind of cgroup event.

**Comment by jangras** [comment]
seeing the same messages:

Docker version: 18.06.0-ce
Kernel: 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:15 UTC 2017 x86_64 x86_64 x86_64 
k8s: v1.13.2

**Comment by CMajeri** [comment]
Hello,

I hit this head on.
It doesn't seem to be an issue with docker.
I'm using kubernetes 1.10.3

I've had a look inside the code and it looks like runc is trying to set "Delegate" on a slice.
In the "Create" call ( https://github.com/kubernetes/kubernetes/blob/v1.10.3/pkg/kubelet/cm/cgroup_manager_linux.go#L468 )

kubelet "adapt"s the name to fit systemd's nomenclature, resulting in this call :
https://github.com/kubernetes/kubernetes/blob/v1.10.3/pkg/kubelet/cm/cgroup_manager_linux.go#L93

Which adds "systemdSuffix" to the cgroup name. This is a const set to 1
https://github.com/kubernetes/kubernetes/blob/v1.10.3/pkg/kubelet/cm/cgroup_manager_linux.go#L49

When this is passed to runc, the "UseSystemd()" call works because it tests "hasDelegate" only on scope units (in lower versions of runc, before commit 0e16bd9b53eb3c57ea6fe59fc6d9385c2edb9fd9) https://github.com/opencontainers/runc/blob/3cbb2fa3c4e8f307817317c455268c6e58703e0b/libcontainer/cgroups/systemd/apply_systemd.go#L153 and doesn't check for slice.

And then during the Apply() call from the ContainerManager's Create(), since we set the "slice" suffix on the cgroup name, everything breaks and the ContainerManager can't be created anymore.

So I guess we need to "not use" that version of kubernetes, and instead find one that's been built with a more recent runc library. So, update kubernetes would be my best guess for a fix.

**Comment by miry** [comment]
same issue for 1.14.0-alpha3

**Comment by hchenxa** [comment]
I also find this issue with kubernete 1.12.4 when using systemd as cgroup-driver, after the kubelet started, this will flood my syslog, but after stop the kubelet, the message only shown when operate the docker container like create/destroy docker container.

**Comment by oiooj** [comment]
Same issue:

```
Docker version: 18.09.2-ce
Linux 4.20.0-1.el7.elrepo.x86_64 #1 SMP Sun Dec 23 20:11:51 EST 2018 x86_64 x86_64 x86_64 GNU/Linux
k8s: v1.12.4
```

**Comment by gocoolcat** [comment]
Seeing the same messages:

Docker Version: 17.12.1-ce
Kernel: 3.10.0-693.43.1.el7.x86_64 #‌1 SMP Thu Oct 11 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
k8s: v1.10.0

I am wondering what the consequence of this is. Is it just message flooding (harmless) or can it cause any problems ?

**Comment by dmanbu** [comment]
Seeing the same messages:

k8s: v1.14.1
Docker Version: 18.09.4-ce
Kernel: 3.10.0-862.el7.x86_64 #1 SMP Fri Apr 20 16:44:24 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

cgroup-driver=systemd

**Comment by jguitard** [comment]
Seeing the same messages:

k8s:  v1.12.3
Docker version: 18.06.1-ce, build e68fc7a
Kernel: 4.9.0-8-amd64 #1 SMP Debian 4.9.144-3.1 (2019-02-19) x86_64 GNU/Linux
OS: Debian GNU/Linux 9 (stretch)
Hardware configuration: QEMU/KVM (baremetal)

**Comment by wiwiwa** [comment]
Same issue

k8s: v1.14.1
Docker Version: 18.09.4-ce
Linux k8s-2 5.0.5-arch1-1-ARCH #1 SMP PREEMPT Wed Mar 27 17:53:10 UTC 2019 x86_64 GNU/LinuxGNU/Linux

cgroup-driver: systemd

**Comment by zhtaoit** [comment]
Same issue
K8s: v1.14.0
Docker：18.06.3-ce
Linux：4.4.178-1.el7.elrepo.x86_64
cgroup-driver: systemd

**Comment by Tyrion85** [comment]
Same issue

k8s: 1.14.1
Docker: 18.09.5
Linux: CentOS Linux release 7.6.1810 (Core) 
Linux version 3.10.0-957.10.1.el7.x86_64 (mockbuild@kbuilder.bsys.centos.org) (gcc version 4.8.5 20150623 (Red Hat 4.8.5-36) (GCC) ) #1 SMP Mon Mar 18 15:06:45 UTC 2019
cgroup-driver: systemd

**Comment by hchenxa** [comment]
not sure if this have any function impact, I just disable the message by configuring the rsyslog:
```
[root@zen-ec-342-master-1 rsyslog.d]# pwd
/etc/rsyslog.d
 
[root@zen-ec-342-master-1 rsyslog.d]# cat ignore-systemd-session-slice.conf
if ($programname == "systemd") and ($msg contains "_systemd_test_default.slice" or$msg contains "systemd-test-default-dependencies.scope") then {
  stop
}
```

**Comment by HaveFun83** [comment]
Same here:
k8s: 1.14.1
Docker: 18.09.5
Linux: Debian 9.9
cgroup-driver: systemd

**Comment by peters0908** [comment]
Same here:
k8s: 1.13.4
Docker: 18.09.3
Linux: RHEL 7.6 3.10.0
cgroup-driver: systemd

**Comment by terryxychan** [comment]
Same here:
k8s: v1.15.0
Docker: 18.06.2-ce
Linux: Ubuntu 16.04
cgroup-driver: systemd

**Comment by yanmaipian** [comment]
docker:18.06.2-ce
k8s:v1.15.3
linux:CentOS Linux release 7.6.1810 (Core)
cgroup-driver: systemd

**Comment by wdec** [comment]
docker:18.09.2
k8s:v1.15.0
linux:Ubuntu 18.04.1 LTS
cgroup-driver: systemd

**Comment by plop28** [comment]
Same here :
docker 18.09.1
linux: Debian 10.1
cgroup-driver: systemd

**Comment by weathery** [comment]
The same problem.
``` Shell
# cat /etc/centos-release
CentOS Linux release 7.7.1908 (Core)

# uname -a
Linux vm02-k8s-node-2 3.10.0-1062.4.1.el7.x86_64 #1 SMP Fri Oct 18 17:15:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

# docker version
Client:
 Version:           18.09.9
 API version:       1.39
 Go version:        go1.11.13
 Git commit:        039a7df9ba
 Built:             Wed Sep  4 16:51:21 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.9
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.11.13
  Git commit:       039a7df
  Built:            Wed Sep  4 16:22:32 2019
  OS/Arch:          linux/amd64
  Experimental:     false

# docker info | grep -i cgroup
Cgroup Driver: systemd

# kubectl get nodes -A
NAME              STATUS   ROLES    AGE   VERSION
vm01-k8s-01       Ready    master   46d   v1.14.8
vm01-k8s-02       Ready    master   46d   v1.14.8
vm01-k8s-03       Ready    master   46d   v1.14.8
vm02-k8s-node-1   Ready    <none>   34d   v1.14.8
vm02-k8s-node-2   Ready    <none>   34d   v1.14.8

# rpm -qa|grep containerd     
containerd.io-1.2.6-3.3.el7.x86_64
```

**Comment by TomaszKlosinski** [comment]
Same story here:
k8s: v1.16.3
docker: 18.06.2-ce
linux: Ubuntu 18.04.3 LTS
cgroup-driver: systemd

**Comment by csallen1204** [comment]
Same Issue Here:
k8s: v1.16.2
docker: 18.06.3-ce
linux: Centos 7.6
cgroup-driver: systemd

**Comment by xinghui-hou** [comment]
Same Issue Here:
k8s: v1.15.5
docker: 18.6.2
linux: Centos 7.4
kernel: 4.4.200
cgroup-driver: systemd

**Comment by gothicfann** [comment]
Same Issue Here:
k8s: v1.15.5
docker: 18.06.2-ce
os: CentOS 7.7
kernel: 3.10.0-1062.1.2
cgroup-driver: systemd

**Comment by Reamer** [comment]
Seems to be fixed on CentOS
k8s: Openshift 3.11
docker: docker-common-1.13.1-109.gitcccb291.el7.centos.x86_64
os: CentOS 7.7
kernel: 3.10.0-1062.12.1.el7.x86_64
Cgroup Driver: systemd

**Comment by niklasember** [comment]
> Seems to be fixed on CentOS
> k8s: Openshift 3.11
> docker: docker-common-1.13.1-109.gitcccb291.el7.centos.x86_64
> os: CentOS 7.7
> kernel: 3.10.0-1062.12.1.el7.x86_64
> Cgroup Driver: systemd

Same here, can't find any log entries about it anymore.

k8s: v1.16.2
docker: 18.06.1-ce
os: Centos 7.7.1908
kernel: 3.10.0-1062.4.1.el7.x86_64
cgroup driver: systemd

**Comment by guikcd** [comment]
Exemple temporary fix is to disable logging these kind of messages: https://www.ibm.com/support/pages/recurring-messages-complain-scope-libcontainer-nnnnn-has-no-pids-refusing

**Comment by rdxmb** [comment]
Can somebody test with me if this is fixed with newer Docker-Versions?

 >  Mrunal Patel 2018-06-18 14:35:45 UTC
> This should be fixed with the latest docker build.

https://bugzilla.redhat.com/show_bug.cgi?id=1584759

In my nodes with docker 18.06.3-ce I have the problem, on a nodes with docker 19.03.13  I do not have the problem.

**Comment by rdxmb** [comment]
Can reproduce: Problem seems solved with docker 19.03.13

So can this issue be closed?

**Comment by ep4eg** [comment]
The bug was fixed in runc [1.0.0-rc7+](https://github.com/opencontainers/runc/commit/cd41feb46b939b6179fbc3c388e33788f72ca38c#diff-d29b03d554c55e3f7c83497bc0f1d4aa2648660fb0e67ae2803efc0dc1915594), so the easiest way to get rid of those messages is to update Docker to 19.0.3+
/lifecycle stale
